soroban-rpc: Add LedgerEntry SQLite storage

2opremio opened the pull request:
Initially migrated from https://github.com/stellar/go/pull/4698 and then adapted to use an on-disk SQLite DB

2opremio commented:
CC @tamirms

2opremio commented:
Depends on https://github.com/stellar/go/pull/4715

tsachiherman commented on cmd/soroban-rpc/internal/daemon/daemon.go:
```diff
@@ -19,6 +24,53 @@ func Start(cfg config.LocalConfig) (exitCode int) {
+	captiveCoreTomlParams.Strict = true
+	captiveCoreToml, err := ledgerbackend.NewCaptiveCoreTomlFromFile(cfg.CaptiveCoreConfigPath, captiveCoreTomlParams)
+	if err != nil {
+		logger.WithError(err).Fatal("Invalid captive core toml")
```
I think that you might want to exit here; the sooban-rpc wouldn't be operative without a captive core.

2opremio replied:
It's a fatal log, it will exit

tsachiherman commented on cmd/soroban-rpc/internal/ledgerentry_storage/db.go:
```diff
@@ -0,0 +1,295 @@
+
+	err = runMigrations(ret.db.DB, "sqlite3")
+
+	return ret, err
```
I'd suggest that we'll be more atomic; if err != nil, do not return ret. ( and close the db )

tsachiherman commented on cmd/soroban-rpc/internal/ledgerentry_storage/db.go:
```diff
@@ -0,0 +1,295 @@
+	ledgerEntryBase64 := results[0]
+	var result xdr.LedgerEntry
+	err = xdr.SafeUnmarshalBase64(ledgerEntryBase64, &result)
+	return result, err
```
same comment as before: if err != nil, return the empty `xdr.LedgerEntry`

tsachiherman commented on cmd/soroban-rpc/internal/ledgerentry_storage/db.go:
```diff
@@ -0,0 +1,295 @@
+	return result, err
+}
+
+func flushLedgerEntryBatch(tx *sqlx.Tx, encodedKeyEntries map[string]*string) error {
```
could you modify the encodedKeyEntries to be an array rather then a map ?
while I do like maps, iteration over maps yields unpredictable order. That means that the entries in the database would have different ordering for the same data. ideally, this would not be the case.

2opremio replied:
A map is used to coalesce multiple modifications to the same key (the same key can be modified multiple times or deleted). You cannot do that with a slice.

2opremio replied:
Also, with a slice you cannot batch queries since updates and deletions would need to be done sequentially and in order.

tsachiherman replied:
sure; we can use a data structure like the following, which would allow us to be persistent:
```golang
type encodedKeyEntry struct {
  key string
  val *string
}
type encodedKeyEntries struct {
   keys []encodedKeyEntry
   keysMaps map[string]uint // translation of key -> index in keys array
}
```

tsachiherman replied:
as for the batching : 
for update from archives, there are no deletions, so it's very much consistent.
when we write ledger entries, we always want to batch these on ledger boundary ( which would guarantee the same outcome.. )

2opremio replied:
I think this overcomplicates things just to ensure a fixed traversal ordering (which isn’t important for the queries we use)

tsachiherman commented on cmd/soroban-rpc/internal/ledgerentry_storage/db.go:
```diff
@@ -0,0 +1,295 @@
+		if err != nil {
+			return err
+		}
+		_, err = tx.Exec(sqlStr, args...)
```
duplicated line ?

2opremio replied:
Deletions and updates are queried separately

sreuland commented on cmd/soroban-rpc/internal/daemon/daemon.go:
```diff
@@ -19,6 +24,53 @@ func Start(cfg config.LocalConfig) (exitCode int) {
+	captiveCoreTomlParams.HistoryArchiveURLs = cfg.HistoryArchiveURLs
+	captiveCoreTomlParams.NetworkPassphrase = cfg.NetworkPassphrase
+	captiveCoreTomlParams.Strict = true
+	captiveCoreToml, err := ledgerbackend.NewCaptiveCoreTomlFromFile(cfg.CaptiveCoreConfigPath, captiveCoreTomlParams)
```
@SirTyson , has [a PR merging soon in go repo](https://github.com/stellar/go/pull/4733/files#diff-c7702ef819e52917ed3868588dff5273774c61a6aa8d36c1644d7ba251fb7756) which will config captive core to use BucketListDB by default, no change needed here, just wanted to mention it, as once it's in, might help in some of the spin up time.

2opremio replied:
Ah, fantastic. I will bump the PR and check the spin up time!

tsachiherman commented on cmd/soroban-rpc/internal/ledgerentry_storage/ledgerentry_storage.go:
```diff
@@ -0,0 +1,238 @@
+package ledgerentry_storage
```
how do you feel about naming the packages `data/ledger` and `data/events` ?

2opremio replied:
I think that the extra level of indirection is not necessary and `ledgerentry_storage` is much more descriptive than `data/ledger`

tsachiherman commented on cmd/soroban-rpc/internal/ledgerentry_storage/ledgerentry_storage.go:
```diff
@@ -0,0 +1,238 @@
+	"github.com/stellar/go/xdr"
+)
+
+// TODO: Make this configurable?
```
yes - good call. no need to expose that to the UI for now, but having this a configurable parameter makes sense.

tsachiherman commented on cmd/soroban-rpc/internal/ledgerentry_storage/ledgerentry_storage.go:
```diff
@@ -0,0 +1,238 @@
+
+	ls.logger.Infof("Starting processing of checkpoint %d", startCheckpointLedger)
+	// TODO: should we make this configurable?
+	checkpointCtx, cancelCheckpointCtx := context.WithTimeout(ctx, 30*time.Minute)
```
move the `30*time.Minute` to be a configurable paramter.

tsachiherman commented on cmd/soroban-rpc/internal/ledgerentry_storage/ledgerentry_storage.go:
```diff
@@ -0,0 +1,238 @@
+		}
+		if err != nil {
+			// TODO: we probably shouldn't panic, at least in case of timeout
+			panic(err)
```
no need to panic here. Just return the error and let the caller decide what to do next.

2opremio replied:
👍

tsachiherman commented on cmd/soroban-rpc/internal/ledgerentry_storage/ledgerentry_storage.go:
```diff
@@ -0,0 +1,238 @@
+		tx.UpsertLedgerEntry(key, *entry)
+		entryCount++
+
+		if entryCount%10000 == 0 {
```
define the 10000 as a const at the top of the file.

2opremio commented:
@tamirms  Thanks! I took a look and it seems more robust but considerably less efficient than what I am using (both in the implementation and in the fact that it seems to treat creation and updates separately, which u avoid using upserts)

2opremio commented:
> and in the fact that it seems to treat creation and updates separately, which I avoid using upserts

I saw that you can simply treat both creations and updates as upserts.

2opremio commented:
I didn't dig on where the performance difference lies but I also observed that it breaks the `db.go` module abstraction:

* I _compact_ entires at the `ledgerkey->ledgerentry` level within the db module, which is where the compaction makes sense since it's a DB performance detail
*  However, the change compactor does so at the change level, which is a bit odd.

2opremio commented:
PTAL

2opremio commented on cmd/soroban-rpc/internal/ledgerentry_storage/db_test.go:
```diff
@@ -0,0 +1,223 @@
+	return key, entry
+}
+
+func TestConcurrency(t *testing.T) {
```
The test already does that (it checks that there is no value in the DB after the Upsert)

2opremio commented on cmd/soroban-rpc/internal/ledgerentry_storage/ledgerentry_storage.go:
```diff
@@ -0,0 +1,226 @@
+		ls.logger.Infof("Found an empty database, filling it in from the most recent checkpoint (this can take up to 30 minutes, depending on the network)")
+		startCheckpointLedger, err = ls.fillEntriesFromLatestCheckpoint(ctx, archive)
+		if err != nil {
+			panic(err)
```
Yeah, good point, I need to change GoLand's settubgs so that it doesn't hide the error handling.

2opremio replied:
Ah, actually, I cannot do bubble them up because `run()` is executed in a goroutine.

2opremio commented on cmd/soroban-rpc/internal/ledgerentry_storage/db.go:
```diff
@@ -0,0 +1,308 @@
+func flushLedgerEntryBatch(tx *sqlx.Tx, encodedKeyEntries map[string]*string) error {
+	upsertCount := 0
+	upsertSQL := sq.Replace(ledgerEntriesTableName)
+	var deleteKeys = make([]interface{}, 0, len(encodedKeyEntries))
```
Good catch, I didn't know it would cast automagically

2opremio commented on cmd/soroban-rpc/internal/ledgerentry_storage/db_test.go:
```diff
@@ -0,0 +1,223 @@
+}
+
+func NewTestDB() (DB, string) {
+	dbPath := path.Join(os.TempDir(), fmt.Sprintf("%08x.sqlite", rand.Int63()))
```
Unfortunately I cannot use it in the benchmark and it's not worth having two separate version of `NewTestDB()`

2opremio commented:
@tamirms PTAL

2opremio commented:
Ah, I get it now!! Thanks! I will incorporate another test

On Thu, 19 Jan 2023 at 12:24 tamirms ***@***.***> wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In cmd/soroban-rpc/internal/ledgerentry_storage/db_test.go
> <https://github.com/stellar/soroban-tools/pull/322#discussion_r1081130856>
> :
>
> > +		LastModifiedLedgerSeq: 1,
> +		Data: xdr.LedgerEntryData{
> +			Type:         xdr.LedgerEntryTypeContractData,
> +			ContractData: &data,
> +		},
> +		Ext: xdr.LedgerEntryExt{},
> +	}
> +	var key xdr.LedgerKey
> +	err := key.SetContractData(data.ContractId, data.Key)
> +	if err != nil {
> +		panic(err)
> +	}
> +	return key, entry
> +}
> +
> +func TestConcurrency(t *testing.T) {
>
> @2opremio <https://github.com/2opremio> I think there's a subtle
> difference between the test case I describe and the test you wrote. In your
> test you check that the GetLedgerEntry() returns ErrEmptyDB while the
> transaction is ongoing and you also check that GetLedgerEntry() returns
> the updated value after the transaction is committed. But, what I'm asking
> about is when the read only transaction is created in the middle of the
> write transaction and the value is fetched using the read only transaction
> *after* after the write transaction is committed. I think in order to
> test this you wouldn't be able to call GetLedgerEntry() because
> GetLedgerEntry() encapsulates the read only transaction within the
> function. I think you need to create the read only transaction in the test
> and then call getLedgerEntry() to make sure the ledger entry is fetched
> inside the read only transaction but after the write transaction is
> comitted.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/stellar/soroban-tools/pull/322#discussion_r1081130856>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AASA4JGUASCCO2RPK6OVDMLWTEP7HANCNFSM6AAAAAATDRRZBI>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

2opremio commented:
>     1. Mitigate [checkpoint starvation](https://github.com/WiseLibs/better-sqlite3/blob/master/docs/performance.md#checkpoint-starvation)

I was actually looking into this.

When running against pubnet I did notice that the WAL file grows to the same size as the DB file and never goes down in size.

```
fons@studioso soroban-tools % ls -lh soroban_rpc.*
-rw-r--r--  1 fons  staff   9.3G Jan 20 12:28 soroban_rpc.sqlite
-rw-r--r--  1 fons  staff    19M Jan 20 03:47 soroban_rpc.sqlite-shm
-rw-r--r--  1 fons  staff   9.4G Jan 20 12:28 soroban_rpc.sqlite-wal
```

I did see the `wall_checkpoint` pragma, but I don't see how to supply it to the Go SQLite driver, it's not available as an option to the connection string (see https://github.com/mattn/go-sqlite3#connection-string ).
 

>     2. Investigate how soroban-rpc can be deployed using a persistent volume. I anticipate if we do a rolling deployment (where the new version is up before the old version is stopped) we will encounter some errors because sqlite does not allow multiple concurrent writers to the db.

You mean, when used in the quickstart image? 👍

2opremio commented on cmd/soroban-rpc/internal/ledgerentry_storage/db.go:
```diff
@@ -0,0 +1,322 @@
+	return ret, nil
+}
+
+func (l *ledgerUpdaterTx) PrepareContext(ctx context.Context, query string) (*sql.Stmt, error) {
```
It implements an interface used by at:

```
ret.stmtCache = sq.NewStmtCache(tx)
```

2opremio replied:
Ah I get it now. That is true!

2opremio commented:
I have addressed all the comments and also made sure that the WAL file is checkpointed after every write commit.

Ah, while I was at it, I also parallelized the checkpoint DB pre-filling with captive core's preparation

tsachiherman commented on cmd/soroban-rpc/internal/ledgerentry_storage/db.go:
```diff
@@ -0,0 +1,340 @@
+)
+
+//go:embed migrations/*.sql
+var migrations embed.FS
```
I like this one.

2opremio replied:
👍 I got it from @tamirms

tsachiherman commented on cmd/soroban-rpc/internal/ledgerentry_storage/db.go:
```diff
@@ -0,0 +1,340 @@
+
+	ret := &sqlDB{
+		db:                  db,
+		postWriteCommitHook: postWriteCommitHook,
```
I think that implementing this method as a sqlDB method would be cleaner; i.e.
`func (db *sqlDB) PostWriteCommit() error`.

I think that the code above is not safe:
The `db` variable is a pointer; and when using anonymous functions, the compiler attempts to keep track of all the external memory location. in this case the `db` escapes the scope of the function and therefore need to be tracked. However, given that the `db` itself was not allocated as a pointer, I'm not sure if it would always work correctly.

2opremio replied:
The garbage collector will keep track of the pointer just fine.

The idea behind using a hook was to make the implementation DB-independent (all the SQLite-specific stuff is only included in the constructor function)

2opremio commented on cmd/soroban-rpc/internal/ledgerentry_storage/db.go:
```diff
@@ -0,0 +1,340 @@
+	if err := l.tx.Commit(); err != nil {
+		return err
+	}
+	if l.postWriteCommitHook() != nil {
```
Good catch!

sreuland commented on cmd/soroban-rpc/main.go:
```diff
@@ -62,6 +74,37 @@ func main() {
 			},
 			Usage: "minimum log severity (debug, info, warn, error) to log",
 		},
+		&config.ConfigOption{
```
I created separate ticket on quickstart to update it's config on soroban runtime, which will now need to include some of these new settings:
https://github.com/stellar/quickstart/issues/406

quickstart won't be able to run this version of rpc until then.

2opremio replied:
👍 thanks!

2opremio commented:
OK, I think everything is finally addressed. Thanks for all the review rounds!

Integration tests are failing due to an unrelated reason https://github.com/stellar/soroban-tools/pull/344 (I will rebase and get it all green before merging)

2opremio commented on cmd/soroban-rpc/internal/ledgerentry_storage/db.go:
```diff
@@ -255,6 +249,7 @@ func (l *ledgerUpdaterTx) UpsertLedgerEntry(key xdr.LedgerKey, entry xdr.LedgerE
 	// safe since we cast to string right away
```
Good idea, I implemented it
